bson/_cbsonmodule.c

```diff
@@ -597,26 +597,21 @@ static int _write_regex_to_buffer(
         Py_DECREF(encoded_pattern);
         return 0;
     }
-    status = check_string((const unsigned char*)pattern_data,
-                          pattern_length, check_utf8, 1);
-    if (status == NOT_UTF_8) {
-        PyObject* InvalidStringData = _error("InvalidStringData");
-        if (InvalidStringData) {
-            PyErr_SetString(InvalidStringData,
-                            "regex patterns must be valid UTF-8");
-            Py_DECREF(InvalidStringData);
-        }
-        Py_DECREF(encoded_pattern);
-        return 0;
-    } else if (status == HAS_NULL) {
-        PyObject* InvalidDocument = _error("InvalidDocument");
-        if (InvalidDocument) {
-            PyErr_SetString(InvalidDocument,
-                            "regex patterns must not contain the NULL byte");
-            Py_DECREF(InvalidDocument);
+
+    if (check_utf8) {
```

Comment:
As Bernie mentioned above we need to keep the NULL byte check. For BSON keys we perform the same check like this:
https://github.com/mongodb/mongo-python-driver/blob/be3008aa11/bson/_cbsonmodule.c#L1230-L1239